bpf/flows.c

```diff
@@ -325,6 +335,64 @@ static inline int flow_monitor(struct __sk_buff *skb, u8 direction) {
     }
     return TC_ACT_OK;
 }
+static inline int export_packet_payload (struct __sk_buff *skb) {
+    void *data_end = (void *)(long)skb->data_end;
+    void *data = (void *)(long)skb->data;
+    payload_meta meta;
+    struct ethhdr *eth  = data;
+    struct iphdr  *ip;
+    struct udphdr *udp_data;
+    
+
+    if ((void *)eth + sizeof(*eth) > data_end) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    ip = data + sizeof(*eth);
+    if ((void *)ip + sizeof(*ip) > data_end) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    udp_data = (void *)ip + sizeof(*ip);
+    if ((void *)udp_data + sizeof(*udp_data) > data_end) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    if (eth->h_proto != __bpf_htons(ETH_P_IP) && (eth->h_proto != __bpf_htons(ETH_P_IPV6))) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    //Only analyze UDP packets
+    if (ip->protocol != IPPROTO_UDP ) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    __be16 dport = udp_data->dest;
+    __be16 sport = udp_data->source;
+    //TODO: Update port number/filters to be read from ENV variable
+    __be16 portFromPanoFilter = __bpf_htons(53);
+    
+    if (dport == portFromPanoFilter || sport == portFromPanoFilter) {
+        meta.if_index = skb->ifindex;
+        meta.pkt_len = data_end - data;
+        meta.payload = data;
+        bpf_perf_event_output(skb, &packet_payloads, ((u64) meta.pkt_len << 32) | BPF_F_CURRENT_CPU, &meta, sizeof(meta));
```

Comment:
@shach33 So currently you are just sending the packet data pointer to userspace, imo this has a race condition here, if some piece of code calls skb_free over the packet in between ebpf sending it and userspace reading it there is a possibility of null pointer exception (or segmentation fault) in the userspace agent. 
If my understanding is correct can you explain the reasoning for sending the pointer and not deepcopy the packet.

cc @msherif1234 do you think this assertion that there is a race condition here is correct?